### PR TITLE
JavaScript capstone project - API-based web app #PR

### PR DESCRIPTION
**Pull request by Sonikak004**
**👥 Authors: [Sonika K](https://github.com/Sonikak004) and [Alfred M](https://github.com/badger-99)**
-------
### In this PR we have : 👍 

- [x] found an API to base the development of the webapp around it.
    - [Pokeapi](https://pokeapi.co/): data about the Pokémon world.
- [x] Used [Involvement API](https://www.notion.so/microverse/Involvement-API-869e60b5ad104603aa6db59e08150270) to record the user interactions (likes, comments). 
- [x] Followed the layout of the wireframes provided and personalized the rest of the design including colors, typographies, spacings, etc.
- [x] Interfaces: 
    - The home page.
    - The comments popup.
     - [x] Home page:
          - When the page loads, the webapp retrieves data from Pokeapi and uses Involvement API to show the item likes.
          - When the user clicks on the "Comments" button, the Comments popup appears.
          - Home page header, navigation and footer is similar to the given mockup.
     - [x] Comments popup:
          - When the popup loads, the webapp retrieves data from Pokeapi and shows details about the selected item and uses the Involvement API to show the item comments.
          - When the user clicks on the "Comment" button, the data is recorded in the Involvement API and the screen is updated.
- [x] Counters
   - For the number of items (home).
   - For the number of comments (comments popup).
- [x] Tests 
   -  Added tests for the items counter.
   - Added tests for comments counter
- [x] No linter errors
- [x] Correct use of GitFlow

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+};
+
+function updateLikeCount(likesButtonId) {
+  const likesButton = document.getElementById(likesButtonId);
+  if (likesButton) {
+    const pokemonName = likesButton.getAttribute('id');
+    const likeCount = likesData[pokemonName] || 0;
+    likesButton.textContent = likeCount.toString();
+  }
+}
```
Comment (Felix45) [inline src/index.js]: - [ Required ] Good job so far guys, I think it would be better if you turn this function in to an `arrow function` the same way you've done for other functions in this project. Use of `ES6` syntax is one of the project requirements so let's fix this to satisfy the requirements.

## Example code.
```JavaScript
const updateLikeCount = (likesButtonId) => {
  // Your code goes here
}
```

## Project wireframes
Screenshot from 2023-07-08 01-03-16

File: src/index.js [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+      updateLikeCount(likesButtonId); // Update the like count in the UI
+      updateLikeCountInPopup(pokemonName, newLikesCount); // Update the like count in the popup
+      localStorage.setItem('likesData', JSON.stringify(likesData)); // Save the updated likes data in local storage
+      // Update the like count in the button
+      likeCounter(likesButtonId, likesData);
+    }
+  } catch (error) {
+    // Handle errors
+  }
+}
```
Comment (Felix45) [inline src/index.js]: - [ Required ] Good job so far guys, I think it would be better if you turn this function in to an `arrow function` the same way you've done for other functions in this project. Use of `ES6` syntax is one of the project requirements so let's fix this to satisfy the requirements.

## Example code.
```JavaScript
const likePokemon = async (likesButtonId) => {
   try {
    const response = await fetch(`https://us-central1-involvement-api.cloudfunctions.net/capstoneApi/apps/${apiKey}/likes`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ item_id: pokemonName }),
    });
   // Your code goes here
}
```

## Project wireframes
Screenshot from 2023-07-08 01-03-16

File: src/updatelike.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
+function updateLikeCountInPopup(pokemonName, likeCount) {
+  const popupContainer = document.getElementById('container');
+  if (popupContainer) {
+    const likesCountElement = popupContainer.querySelector('#likesCount');
+    if (likesCountElement && likesCountElement.getAttribute('data-pokemon') === pokemonName) {
+      likesCountElement.textContent = likeCount.toString();
+    }
+  }
+}
```
Comment (Felix45) [inline src/updatelike.js]: - [ Required ] Good job so far guys, I think it would be better if you turn this function in to an `arrow function` the same way you've done for other functions in this project. Use of `ES6` syntax is one of the project requirements so let's fix this to satisfy the requirements.

## Example code.
```JavaScript
const updateLikeCountInPopup = (pokemonName, likeCount) => {
  // Your code goes here
}
```

## Project wireframes
Screenshot from 2023-07-08 01-03-16

File: src/likecounter.js [hunk 4]
```diff
@@ -0,0 +1,10 @@
+export default function likeCounter(buttonId, likesData) {
+  const button = document.getElementById(buttonId);
+  if (button) {
+    const pokemonName = button.getAttribute('id');
+    const likeCount = likesData[pokemonName] || 0;
+    button.innerHTML = likeCount.toString();
+    return likeCount;
+  }
+  return 0;
+}
```
Comment (Felix45) [inline src/likecounter.js]: - [ Required ] Good job so far guys, I think it would be better if you turn this function in to an `arrow function` the same way you've done for other functions in this project. Use of `ES6` syntax is one of the project requirements so let's fix this to satisfy the requirements.

## Example code.
```JavaScript
const likeCounter = (buttonId, likesData) => {
  // Your code goes here
}

export default likeCounter;
```

## Project wireframes
Screenshot from 2023-07-08 01-03-16

File: src/commentCounter.test.js [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+describe('testing the comment counting function', () => {
+  test('the comment container returns count of 1 after appending comment element', () => {
+    const container = document.createElement('div');
+    const comment = document.createElement('p');
+    comment.innerText = 'Easy peasy lemon squeazy';
+    container.appendChild(comment);
+    expect(counter(container)).toBe(1);
+  });
+});
```
Comment (Felix45) [inline src/commentCounter.test.js]: - [ Required ] Good job adding tests for the `commentCounter` module, however, I think it would be better if you make sure that you have tests for all edge cases e.g think at least about the scenario in which there are 0 items on the page (counter function should return 0)

## Example code
```JavaScript
describe('testing the comment counting function', () => {
  test('the comment container returns count of 0', () => {
    const container = document.createElement('div');
    const comment = document.createElement('p');
    expect(counter(container)).toBe(0);
  });
});
```

File: src/likecounter.test.js [hunk 6]
```diff
@@ -0,0 +1,31 @@
+/**
+ * @jest-environment jsdom
+ */
+
+import likeCounter from './likecounter.js';
+
+describe('likeCounter', () => {
```
Comment (Felix45) [inline src/likecounter.test.js:7]: - [ Required ] Good job adding tests for the `likescounter` module, however, I think it would be better if you make sure that you have tests for all edge cases e.g think at least about the scenario in which there are 0 likes on an item (counter function should return 0)

## Example code
```JavaScript
/**
 * @jest-environment jsdom
 */

import likeCounter from './likecounter.js';

describe('likeCounter', () => {
  beforeEach(() => {
    document.body.innerHTML = `
      <button id="testButton">Likes: <span id="likeCount">0</span></button>
    `;
  });

  afterEach(() => {
    document.body.innerHTML = '';
  });

  it('likes counter should return zero likes', () => {
    const buttonId = 'testButton';
    const likesData = {
      testButton: 0,
    };

    const button = document.getElementById(buttonId);

    const updatedLikeCount = likeCounter(buttonId, likesData);

    expect(updatedLikeCount).toBe(0);
    expect(button.innerHTML).toBe('0');
  });
});
```

**Review by Felix45 — CHANGES_REQUESTED** [review]
Hi @Sonikak004 & @badger-99, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!



## Highlights :heavy_check_mark: 
- [x] Good UI design :white_check_mark: 
- [x] Git flow is used correctly :white_check_mark: 
- [x] Application works as expected :white_check_mark: 
- [x] Application is integrated well with the Involvement API :white_check_mark: 
- [x] Pull request has a descriptive title and summary :white_check_mark: 
- [x] Repository has a professional readme file :white_check_mark:    


## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by levy002 — APPROVED** [review]
### Hello @team

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



## Highlights
- No linter error :heavy_check_mark: 
- Implemented all suggested changes :heavy_check_mark: 
- Good UI :heavy_check_mark: 

## Status: APPROVED :fire:  :fire: 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).
